stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+ *
+ * @return The value of the n-dimensional integral of \f$f\f$ from \f$a\f$ to
+ \f$b\f$.
+ * @throw std::domain_error no errors will be thrown.
+ */
+
+// hcubature
+template <typename F, typename T_pars, typename T_n, typename T_a, typename T_b,
+          typename T_maxEval, typename T_reqAbsError, typename T_reqRelError>
+return_type_t<T_n, T_a, T_b, T_maxEval, T_reqAbsError, T_reqRelError> hcubature(
+    const F& integrand, const T_pars& pars, const T_n& n, const T_a& a,
+    const T_b& b, const T_maxEval& maxEval, const T_reqAbsError& reqAbsError,
+    const T_reqRelError& reqRelError) {
+  using T_return_type
+      = return_type_t<T_n, T_a, T_b, T_maxEval, T_reqAbsError, T_reqRelError>;
+
+  using T_a_ref = ref_type_t<T_a>;
+  using T_b_ref = ref_type_t<T_b>;
+
+  T_a_ref a_ref = a;
+  T_b_ref b_ref = b;
+
+  scalar_seq_view<T_a_ref> a_vec(a_ref);
+  scalar_seq_view<T_b_ref> b_vec(b_ref);
+
+  internal::one_d out;
+  internal::GenzMalik g;
+
+  if (n == 1) {
+    internal::gauss_kronrod(integrand, a_vec.val(0), b_vec.val(0), out, pars);
+  } else {
+    internal::make_GenzMalik(n, g);
+    internal::integrate_GenzMalik(integrand, g, n, a, b, out, pars);
+  }
+  T_return_type numevals
+      = (n == 1) ? 15 : 1 + 4 * n + 2 * n * (n - 1) + std::pow(2, n);
+  T_return_type evals_per_box = numevals;
+  T_return_type kdiv = out.kdivide;
+  T_return_type err = out.err;
+  T_return_type val = out.result;
+  // convergence test
+  if ((err <= fmax(reqRelError * fabs(val), reqAbsError))
+      || ((maxEval != 0) && (numevals >= maxEval))) {
+    internal::clean_GenzMalik(g);
+    return val;
+  }
+
+  std::multiset<internal::Box> ms;
+  ms.insert(internal::make_box(n, a, b, out));
+
+  while (true) {
+    std::multiset<internal::Box>::iterator it;
+    it = ms.begin();
+    internal::Box box = *it;
+    ms.erase(it);
+    // split along dimension kdiv
+    T_return_type w = (box.b[box.kdiv] - box.a[box.kdiv]) / 2;
+    std::vector<double> ma(box.a);
+
+    ma[box.kdiv] += w;
+    std::vector<double> mb(box.b);
+    mb[box.kdiv] -= w;
+
+    if (n == 1) {
+      internal::gauss_kronrod(integrand, ma[0], box.b[0], out, pars);
+    } else {
+      internal::integrate_GenzMalik(integrand, g, n, ma, box.b, out, pars);
+    }
+    internal::Box box1 = make_box(n, ma, box.b, out);
+    ms.insert(box1);
+
+    if (n == 1) {
+      internal::gauss_kronrod(integrand, box.a[0], mb[0], out, pars);
+    } else {
+      internal::integrate_GenzMalik(integrand, g, n, box.a, mb, out, pars);
+    }
+    internal::Box box2 = make_box(n, box.a, mb, out);
+    ms.insert(box2);
+    val += box1.I + box2.I - box.I;
+    err += box1.E + box2.E - box.E;
+    numevals += 2 * evals_per_box;
+
+    std::vector<double>().swap(box.a);
+    std::vector<double>().swap(box.b);
+    std::vector<double>().swap(ma);
+    std::vector<double>().swap(mb);
+    if (((err <= max(reqRelError * fabs(val), reqAbsError))
+         || ((maxEval != 0) && (numevals >= maxEval)))
```

Comment:
`numevals` records the number of integrand evaluations so far which means that the loop stops _after_  maxEvals has been exceeded. I think it would be better if this condition anticipated the number of evals in the next round and stopped _before_ exceeding maxEvals so that maxEvals actually was a hard upper bound on the total number evals.

I'm also not convinced "`maxEvals=0` means there's no upper bound" is a needed complication; you can always set maxEvals to a number so large it can't be reached within any reasonable time.

Reply:
Comes in next commit.